modules/renter/contractor/contracts.go

```diff
@@ -502,10 +518,39 @@ func (c *Contractor) threadedContractMaintenance() {
 				return
 			}
 			// Perform the actual renew. If the renew fails, return the
-			// contract.
-			newContract, err := c.managedRenew(oldContract, amount, endHeight)
-			if err != nil {
-				c.log.Printf("WARN: failed to renew contract %v: %v\n", id, err)
+			// contract. If the renew fails we check if it has failed before
+			// and remember the height if it hasn't. Once it has failed for a
+			// certain number of blocks in a row and reached its second half of
+			// the renew window, we give up on renewing it and set goodForRenew
+			// to false.
+			newContract, errRenew := c.managedRenew(oldContract, amount, endHeight)
+			if errRenew != nil {
+				// Check if contract has to be replaced.
+				md := oldContract.Metadata()
+				c.mu.RLock()
+				firstRenew, failedBefore := c.firstFailedRenew[md.ID]
+				c.mu.RUnlock()
+				secondHalfOfWindow := blockHeight+allowance.RenewWindow/2 >= md.EndHeight
+				replace := blockHeight-firstRenew >= consecutiveRenewalsBeforeReplacement
```

Comment:
This doesn't mean that you've tried 12 times, it just means that 12 blocks have passed. That might have issues if you come online and start syncing, though I know we don't run this code if we aren't already sync'd.

I think it's probably less fragile though to make the check 12 discreet tries instead of 12 blocks passed since the first failed attempt.

Reply:
It feels less fragile to me because it eliminates the number of external factors that can interfere with the logic here.